src/thumbnailer_slope_optim.cc

```diff
@@ -190,33 +171,38 @@ Thumbnailer::Status Thumbnailer::LossyEncodeSlopeOptim(
 
 Thumbnailer::Status Thumbnailer::LossyEncodeNoSlopeOptim(
     WebPData* const webp_data) {
+  // The webp_data->size and the sum of encoded-frame sizes are inconsistent,
+  // therefore consider the bigger one as the current animation size to ensure
+  // the resulting animation size fit the byte budget.
   int anim_size = 0;
   for (auto& frame : frames_) {
     anim_size += frame.encoded_size;
   }
+  anim_size = std::max(anim_size, int(webp_data->size));
 
-  WebPData new_webp_data;
-  WebPDataInit(&new_webp_data);
+  int num_remaining_frames = frames_.size();
 
   // For each frame, find the best quality value that can produce the higher
   // PSNR than the current one if possible.
   for (auto& frame : frames_) {
     int min_quality = 0;
-    int max_quality = 100;
+    if (!frame.config.lossless) {
+      min_quality = frame.final_quality;
+    }
+    int max_quality = std::min(min_quality + 30, 100);
     frame.config.lossless = 0;
     while (min_quality <= max_quality) {
       int mid_quality = (min_quality + max_quality) / 2;
       int new_size;
       float new_psnr;
       frame.config.quality = mid_quality;
-      if (GetPictureStats(frame.pic, frame.config, &new_size, &new_psnr) !=
-          kOk) {
-        return kStatsError;
-      }
+      CHECK_THUMBNAILER_STATUS(
+          GetPictureStats(frame.pic, frame.config, &new_size, &new_psnr));
 
       if (new_psnr > frame.final_psnr || ((new_psnr == frame.final_psnr) &&
                                           (new_size <= frame.encoded_size))) {
-        if (anim_size - frame.encoded_size + new_size <= byte_budget_) {
+        float extra_budget = (byte_budget_ - anim_size) / num_remaining_frames;
```

Comment:
Are we guaranteed that anim_size is actually smaller byte_budget? Otherwise extra_budget could be negative.

Reply:
I think that `anim_size` is smaller than `byte_budget` in most cases. Might it be better if I write something like : `if (anim_size > byte_budget) return kOk;` [here](https://github.com/googleinterns/step255-2020/blob/b02b1727d9b0135538ce6c3196721cbe1bb11007/src/thumbnailer_slope_optim.cc#L182) in order to guarantee the extra_budget be positive and do nothing otherwise ?

Reply:
Yes that sounds pretty reasonable.